sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/python/BatchGroupUtils.scala

```diff
@@ -0,0 +1,288 @@
+          // the children here
+          BatchGroupedIterator.extractChildren(cbFromPython, output)
+        }
+      } // end of next
+    } // end of Iterator
+  }
+
+}
+
+/**
+ * Runs a `groupby` on the input batch, then split it into separate batches by the grouping
+ * expressions.
+ * Since the rows in the batches are already sorted by Spark, a better performance is probably
+ * achieved in the cudf `groupby`.
+ *
+ * Example Input: (Grouping: `x`)
+ *   A batch of 3 rows, two groups.
+ *         x    y
+ *        --------
+ *         a    1
+ *         b    2
+ *         b    3
+ *
+ * Result:
+ *   Two batches, one group one batch.
+ * First call to next(), returning the batch of group `a`:
+ *         x    y
+ *        --------
+ *         a    1
+ *
+ * Second call to next(), returning the batch of group `b`:
+ *         x    y
+ *        --------
+ *         b    2
+ *         b    3
+ *
+ * Note, the order of the groups returned is NOT always the same with it in the input batch.
+ *
+ * Besides the class does not handle the case of an empty input for simplicity of implementation.
+ * Use the factory to construct a new instance.
+ *
+ * @param input An iterator of batches.
+ * @param inputAttributes The schema of the batch in the `input` iterator.
+ * @param groupingIndices The set of column indices used to do grouping.
+ */
+private[python] class BatchGroupedIterator private(
+    input: Iterator[ColumnarBatch],
+    inputAttributes: Seq[Attribute],
+    groupingIndices: Seq[Int],
+    spillCallback: RapidsBuffer.SpillCallback) extends Iterator[ColumnarBatch] with Arm {
+
+  private val batchesQueue: mutable.Queue[SpillableColumnarBatch] = mutable.Queue.empty
+
+  // Suppose runs inside a task context.
+  TaskContext.get().addTaskCompletionListener[Unit]{ _ =>
+    batchesQueue.foreach(_.close())
+    batchesQueue.clear()
+  }
+
+  override def hasNext: Boolean = batchesQueue.nonEmpty || input.hasNext
+
+  override def next(): ColumnarBatch = {
+    if (groupingIndices.isEmpty) {
+      // Empty grouping expression, so the entire batch is a single group.
+      // Returns it directly.
+      input.next()
+    } else {
+      // Grouping expression is not empty, so returns the first batch in the queue.
+      // If the queue is empty, tries to read and split the next batch from input.
+      if (batchesQueue.isEmpty) {
+        val batch = input.next()
+
+        // Splits the batch per grouping expressions
+        val groupTables = withResource(batch) { b =>
+          withResource(GpuColumnVector.from(b)) { table =>
+            // In Spark, the rows in a batch are already sorted by grouping keys in
+            // the order of `Ascending & NullsFirst` for Pandas UDF plans.
+            // So passes the info to cudf for better performance on `groupBy` operation.
+            val builder = cudf.GroupByOptions.builder()
+            builder.withIgnoreNullKeys(false)
+                   .withKeysSorted(true)
```

Comment:
To my understanding, this should be already ensured by overriding the `requiredChildOrdering` in each Pandas UDF plan.

For example in the `FlatMapGroupsInPandasExec`, it is
```scala
  override def requiredChildOrdering: Seq[Seq[SortOrder]] =
    Seq(groupingAttributes.map(SortOrder(_, Ascending)))
```

Reply:
But it is not obvious that `BatchGroupUtils` is only used by `FlatMapGroupsInPandasExec`, or that it will only be used by it in the future. If we create a new API, even a semi private one, we should document the requirements in the API at a minimum.

Reply:
BatchGroupedIterator is private, but the sorted requirement is documented here.  It would be nice to mention it as a requirement in `executePython`

Reply:
Got it.
The order is required by the `BatchGroupedIterator `, so I added the requirement in its doc.
```scala
 * Runs a `groupby` on the input batch where the rows should be presorted in the order of
 * `Ascending & NullsFirst`, then split it into separate batches by the grouping expressions.
```

Reply:
I would have preferred this on the `executePython` API, because it is the "public" entry point. But this is fine.